datafusion-cli/src/exec.rs

```diff
@@ -249,8 +255,21 @@ pub(super) async fn exec_and_print(
         } else {
             // Bounded stream; collected results are printed after all input consumed.
             let schema = physical_plan.schema();
-            let results = collect(physical_plan, task_ctx.clone()).await?;
+            let mut stream = execute_stream(physical_plan, task_ctx.clone())?;
+            let mut results = vec![];
+            while let Some(batch) = stream.next().await {
+                let batch = batch?;
+                reservation.try_grow(get_record_batch_memory_size(&batch))?;
+                results.push(batch);
+                if let MaxRows::Limited(max_rows) = print_options.maxrows {
+                    // Stop collecting results if the number of rows exceeds the limit
+                    if results.len() >= max_rows && print_options.stop_after_max_rows {
```

Comment:
If you break here it will stop the plan early -- then you need to add `stop_after_max_rows` to permit the plan to run to completion, however, if that option is set then the data is buffered again (which from a memory perspective is the same as increasing the max_rows)

Rather than `break`ing here, I recommend just ignoring the batches after `max_rows` has been reached (aka don't `push` them into `results`) -- I think that would be the most intuitive behavior:
1. Queries would still always run to completion
2. Max rows would control how much memory was used by datafusion-cli buffering

Reply:
Thank you @alamb for review and the great idea! It makes sense to me, addressed in latest PR.